### ISSUE TITLE
Optional path segments

### ISSUE THREAD

**Issue by carbotaniuman** [http4k/http4k#558]
In some other frameworks, it's possible to have an optional placeholder that can be ignored if necessary. Is the router in http4k capable of doing this?

**Comment by daviddenton** [comment]
This isn't really possible because of the way the routing is designed. You can work around it to some extent - I've had a play and this is the best I could come up with - which is adding the duplicated handler

```kotlin
fun main() {
    val handler = { req: Request ->
        val path = req.path("foo") ?: "default"
        Response(Status.OK)
    }

    GET bind routes(
        "/{name}" bind handler,
        "/" bind handler
    )
}

private infix fun Method.bind(rhh: RoutingHttpHandler) = { r: Request -> r.method == this }.asRouter() bind rhh
```

**Comment by carbotaniuman** [comment]
Hmm, the same way the new router in 4.0.0.0 has the ability to filter on body and headers, is it possible to add something that filters on the content of a placeholder `{name}`?

**Comment by carbotaniuman** [comment]
```
fun path(name: String, predicate: (String) -> Boolean) = {req: Request -> req.path(name).let {
    println(it)
    if(it == null) {
        false
    } else {
        predicate(it)
    }
} }.asRouter()
```

I tried this but the path segment doesn't seem to get bound until after the sub router is called.

**Comment by s4nchez** [comment]
Yes, unfortunately, the request is not modified to include the routing information during the matching process. I'm inclined to say we shouldn't change that behaviour, mainly because giving routers the possibility to change the original request can make routing even more complicated.

Happy to consider simpler solutions if anyone has one though.
